pbft/pbft.cpp

```diff
@@ -1041,13 +1056,443 @@ pbft::get_high_water_mark()
+    {
+        return sender== address.uuid;
+    }) != this->current_peers().end();
+}
+
+std::map<bzn::checkpoint_t, std::set<bzn::uuid_t>>
+pbft::validate_and_extract_checkpoint_hashes(const pbft_msg &viewchange_message) const
+{
+    std::map<bzn::checkpoint_t , std::set<bzn::uuid_t>> checkpoint_hashes;
+    for (size_t i{0}; i < static_cast<uint64_t>(viewchange_message.checkpoint_messages_size()); ++i)
+    {
+        bzn_envelope envelope{viewchange_message.checkpoint_messages(i)};
+        pbft_msg checkpoint_message;
+
+        if (!this->crypto->verify(envelope) || !this->is_peer(envelope.sender()) || !checkpoint_message.ParseFromString(envelope.pbft()))
+        {
+            LOG (error) << "Checkpoint validation failure - unable to verify envelope";
+            continue;
+        }
+        checkpoint_hashes[checkpoint_t(checkpoint_message.sequence(), checkpoint_message.state_hash())].insert(envelope.sender());
+    }
+
+    //  filter checkpoint_hashes to only those pairs (checkpoint, senders) such that |senders| >= 2f+1
+    std::map<bzn::checkpoint_t , std::set<bzn::uuid_t>> retval;
+    std::copy_if(checkpoint_hashes.begin(), checkpoint_hashes.end(), std::inserter(retval, retval.end()), [&](const auto& h) {return h.second.size() >= 2 * this->max_faulty_nodes() + 1;});
+
+    return retval;
+}
+
+bool
+pbft::is_valid_viewchange_message(const pbft_msg& viewchange_message, const bzn_envelope& original_msg) const
+{
+    if (!this->is_peer(original_msg.sender()))// TODO: Rich - If redundant, keep this check, remove the other
+    {
+        LOG(debug) << "is_valid_viewchange_message - message not from a peer";
+        return false;
+    }
+
+    if (!(viewchange_message.view() > this->get_view()))
+    {
+        LOG(error) << "is_valid_viewchange_message - new view " <<  viewchange_message.view() << " is not greater than current view " << this->get_view();
+        return false;
+    }
+
+    auto valid_checkpoint_hashes = this->validate_and_extract_checkpoint_hashes(viewchange_message);
+
+    if (valid_checkpoint_hashes.empty())
+    {
+        LOG(error) << "is_valid_viewchange_message - the checkpoint is invalid";
+        return false;
+    }
+    const auto valid_checkpoint = *(valid_checkpoint_hashes.begin());
+
+
+    // TODO: (see KEP-882) Refactor this block into a new method - Isabel: should extract a validate_prepared_proof method
+    // all the the prepared proofs are valid  (contains a pre prepare and 2f+1 mathcin prepares)
+    for (int i{0}; i < viewchange_message.prepared_proofs_size(); ++i)
+    {
+        const bzn_envelope& pre_prepare_envelope{viewchange_message.prepared_proofs(i).pre_prepare()};
+        if (!this->is_peer(pre_prepare_envelope.sender()) || !this->crypto->verify(pre_prepare_envelope))
+        {
+            LOG(error) << "is_valid_viewchange_message - a pre prepare message has a bad envelope, or the sender is not in the peers list";
+            return false;
+        }
+
```

Comment:
check that the preprepare has the right type (preprepare_msg.type() == PBFT_MSG_PREPREPARE iirc)